### DISCUSSION TITLE
Map loading android 11

### DISCUSSION THREAD

**Discussion by ocsike** [mapsforge/vtm#949] [Q&A] [answered]
Hi, 
I want to load maps from a directory, but no luck.  I have write rights too.  The maps are white. I use following code:
` files.add("hungary.map"); //arraylist with required map names to show
                for (DocumentFile dm: mmapfiles) {
                   // MyLog.log(MapActivity.class,"Uri_K1:"+dm.getName()+files.size());
                    if(files.indexOf(dm.getName())>-1){  //if must show
                    MyLog.log(MapActivity.class,"map_uri:"+dm.getUri());
                        try {
                            MapFileTileSource mapFileTileSource = new MapFileTileSource();
                            FileInputStream fis = (FileInputStream) getContentResolver().openInputStream(dm.getUri()); // new FileInputStream(getContentResolver().openFileDescriptor(ii.getData(), "r").getFileDescriptor());
                            mapFileTileSource.setMapFileInputStream(fis);
                            mMapFiles.add(dm); // arraylist for show mapinfo1
                            TileSource.OpenResult openResult = mapFileTileSource.open();
                            MyLog.log(MapActivity.class,"openresult:"+openResult.toString());

                            if(openResult.isSuccess()) {
                                boolean added=  tileSource.add(mapFileTileSource);
                                mMapInfos.add(mapFileTileSource.getMapInfo());  // arraylist for show mapinfo2
                                MyLog.log(MapActivity.class,"map_OK:" + added+", file:"+ dm.getName()+",,"+mMapInfos.size());
                            }
                            else {
                                Toast.makeText(MapActivity.this, openResult.getErrorMessage(), Toast.LENGTH_LONG);
                                MyLog.log(MapActivity.class,"map_error:"+openResult.getErrorMessage()  );
                            }
                            mapFileTileSource.close();
                        }
                        catch (Exception e){
                            MyLog.log(MapActivity.class,"other error:"+e.getMessage());
                        }
                    }
                }
tileSource.setPreferredLanguage("hu");               
                mVectorLayer = mMapView.map().setBaseMap(tileSource);
                mMap.setTheme(VtmThemes.DEFAULT);`
`

`
The log is:
` mao count:11
map_uri:content://com.android.externalstorage.documents/tree/primary%3AGmaps/document/primary%3AGmaps%2FMaps%2Fhungary.map
openresult:FileOpenResult [success=true, errorMessage=null]
map_OK:true, file:hungary.map,,1
`
If I try a sample MapsforgeActivity then map load correctly, but, uri is different:
`content://com.android.externalstorage.documents/document/primary%3AGmaps%2FMaps%2Fhungary.map`

The different content uri cause the problem? Thanks

**Comment by devemux86** [comment, marked as answer]
You have to use the Android storage access framework to grant you read access permission.

See the [GettingStarted](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java) example.

**Reply by ocsike** [reply]
Gettingstarted example is show white empty map for me. The sample MapsforgeActivity is ok.

**Comment by devemux86** [comment]
> Gettingstarted example is show white empty map for me.

Study carefully the GettingStarted example:
```java
 * You'll need a map with filename berlin.map from download.mapsforge.org in device storage.
```
```java
// Note: this map position is specific to Berlin area
mapView.map().setMapPosition(52.517037, 13.38886, 1 << 12);
```

**Reply by ocsike** [reply]
Ok, but if the map is not loaded correctly the mapFileTileSource.open() result is why true?

**Comment by devemux86** [comment]
> if the map is not loaded correctly the mapFileTileSource.open() result is why true?

Without more details, we cannot understand what you mean.

**Comment by ocsike** [comment]
Ok, I do this:
1. Open map with filepicker
`Intent iintent = new Intent(Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT ? Intent.ACTION_OPEN_DOCUMENT : Intent.ACTION_GET_CONTENT);
                            iintent.addCategory(Intent.CATEGORY_OPENABLE);
                            iintent.addFlags(
   Intent.FLAG_GRANT_PREFIX_URI_PERMISSION);

                            iintent.setType("*/*");
                            startActivityForResult(iintent, SELECT_MAP_FILE);`
2. In onActivityResult save the map URI  (with Uri.toString() method ) and show the map. It's work fine.
`try {
                Uri uri = data.getData();
                
                grantUriPermission(getPackageName(), uri, Intent.FLAG_GRANT_READ_URI_PERMISSION);

                final int takeFlags = data.getFlags() & (Intent.FLAG_GRANT_READ_URI_PERMISSION);
                getContentResolver().takePersistableUriPermission(uri, takeFlags);
  
                loadMap(uri);

            }
            catch (Exception e){
                MyLog.log(MapActivity.class,"Gixer:"+e.getMessage());
            }`
`private void loadMap(Uri mapUri){
       try {
           MapFileTileSource tileSource = new MapFileTileSource();
           FileInputStream fis = (FileInputStream) getContentResolver().openInputStream(mapUri);
           tileSource.setMapFileInputStream(fis);
           MyLog.log(MapActivity.class,"map_uri:" + mapUri);
           mVectorLayer = mMapView.map().setBaseMap(tileSource);
           mMap.setTheme(VtmThemes.DEFAULT);
       }
       catch (Exception e){
           MyLog.log(MapActivity.class,"error:" + e.getMessage());
       }
   }`

3. Close application,.
4. Start application again
5. Get the previously saved map uri and load map. It not works.
6. In the Androidmanifest file I defined this too:
` <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"/>
           <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE"/>`

Both case (2,5)  the TileSource.OpenResult result is ok and no error message.

**Comment by devemux86** [comment]
Map library can read `FileInputStream`, so it works with Android storage access framework.

_Android storage access framework implementation is the responsibility of the apps._

---

Study the Android documentation:
- [Open a file](https://developer.android.com/training/data-storage/shared/documents-files#open-file)
- [Persist permissions](https://developer.android.com/training/data-storage/shared/documents-files#persist-permissions)
- Old storage permissions are no longer used in [AndroidManifest](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/AndroidManifest.xml)
- [GettingStarted](https://github.com/mapsforge/vtm/blob/master/vtm-android-example/src/org/oscim/android/test/GettingStarted.java) example is for demonstration purposes

Android documentation:
- https://developer.android.com/training/data-storage/shared/documents-files

---

Your code is not readable, you should use Java code blocks:
`` ```java ``
`` ``` ``
